std/net/isemail.d

```diff
@@ -1863,3 +1857,95 @@ const(T)[] get (T) (const(T)[] str, size_t index, dchar c)
     assert("abc".get(1, 'b') == "b");
     assert("löv".get(1, 'ö') == "ö");
 }
+
+/+
+Replacement for:
+---
+static fourChars = ctRegex!(`^[0-9A-Fa-f]{0,4}$`.to!(const(Char)[]));
+...
+a => a.matchFirst(fourChars).empty)
+---
++/
+private static bool isUpToFourHexChars(Char)(scope const(Char)[] s)
+{
+    import std.ascii : isHexDigit;
+    if (s.length > 4) return false;
+    foreach (i; 0 .. s.length)
+        if (!isHexDigit(s[i])) return false;
```

Comment:
```D
foreach (c; s)
    if (!isHexDigit(c)) return false;
```

Reply:
Although that is nicer syntax, it would auto-decode which is not necessary (all characters I'm searching for are <= 0x7F) and would be a performance hit.

Reply:
`foreach (c; s)` doesn't do auto-decoding.

Reply:
It does if `c` is `dchar` and `s` is a narrow string.

Reply:
My mistake: I thought `c` would default to `dchar` unless given a specific type. Changed to use `foreach`.

Reply:
Are you sure? https://run.dlang.io/is/UGY8jP

Reply:
> I thought c would default to dchar unless given a specific type.

This is a common misconception -- because Phobos treats arrays of `char` weirdly. The language doesn't, it's still an array to the language.